### PR TITLE
9.0 FEATURE: Persist NodeTypeName in DocumentUriPathProjection

### PR DESCRIPTION
**Pull request by dlubitz**
Adds an additional field for the NodeTypeName of the NodeAggregate. This is needed for further processing of the DocumentNodeInfo for redirect handling.

The Neos.RedirectHandler.NeosAdapter provides the feature to disallow the creation of redirect for given NodeTypes. To support this feature also in Neos 9 with the new ContentRepository, the DocumentUriPathProjection needs to provide this information, as we fully rely on this projection data for the new redirect handling.

Relates to: https://github.com/neos/neos-development-collection/issues/4149

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/Feature/RootNodeCreation/Event/RootNodeAggregateDimensionsWereUpdated.php [hunk 1]
```diff
@@ -35,8 +36,9 @@ final class RootNodeAggregateDimensionsWereUpdated implements
     public function __construct(
         public readonly ContentStreamId $contentStreamId,
         public readonly NodeAggregateId $nodeAggregateId,
+        public readonly NodeTypeName $nodeTypeName,
```
Comment (skurfuerst) [inline Neos.ContentRepository.Core/Classes/Feature/RootNodeCreation/Event/RootNodeAggregateDimensionsWereUpdated.php]: we cannot update the event payload here - such a change is very breaking. Why do you need this? You can simply copy the node type name from the existing root(s) :)

File: Neos.ContentRepository.Core/Classes/Feature/RootNodeCreation/Event/RootNodeAggregateDimensionsWereUpdated.php [hunk 2]
```diff
@@ -64,6 +67,7 @@ public static function fromArray(array $values): self
         return new self(
             ContentStreamId::fromString($values['contentStreamId']),
             NodeAggregateId::fromString($values['nodeAggregateId']),
+            NodeTypeName::fromString($values['nodeTypeName'] ?? 'Neos.Neos:Sites'),
```
Comment (skurfuerst) [inline Neos.ContentRepository.Core/Classes/Feature/RootNodeCreation/Event/RootNodeAggregateDimensionsWereUpdated.php]: the CR does not depend on Neos.Neos:Sites

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 3]
```diff
@@ -248,6 +249,7 @@ private function whenRootNodeAggregateDimensionsWereUpdated(RootNodeAggregateDim
                 'nodeAggregateIdPath' => $event->nodeAggregateId->value,
                 'dimensionSpacePointHash' => $dimensionSpacePoint->hash,
                 'nodeAggregateId' => $event->nodeAggregateId->value,
+                'nodeTypeName' => $event->nodeTypeName->value,
```
Comment (skurfuerst) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: before deleting, I'd read the node type name from the projection here, so this way you do not need it in nthe event.

**Review by skurfuerst — CHANGES_REQUESTED** [review]
hey, thanks for your change :) I added some comments to make it less breaking.

all the best,
Sebastian

**Comment by dlubitz** [comment]
Thanks @skurfuerst, I did the changes you suggested.

**Comment by skurfuerst** [comment]
Awesome! I will review again today around noon.

**Comment by bwaidelich** [comment]
Just looking at this briefly without going into the details, I failed to understand why this is needed?
Could you add a little explaining phrase to the description? (I fully trust you on this one ofc, but I'm curious and others might be too)

**Comment by dlubitz** [comment]
> Just looking at this briefly without going into the details, I failed to understand why this is needed? Could you add a little explaining phrase to the description? (I fully trust you on this one ofc, but I'm curious and others might be too)

A added some more background information to the description.

**Comment by skurfuerst** [comment]
looks all good, we should only extend the tests at https://github.com/neos/neos-development-collection/tree/9.0/Neos.Neos/Tests/Behavior/Features/FrontendRouting as well. Do you want to do it in this PR or in a separate one? :) 

All the best,
Sebastian

**Review by skurfuerst — APPROVED** [review]

**Comment by dlubitz** [comment]
@skurfuerst Sure, I'll add some tests. Shall I extend the existing ones? Or shall we add dedicated tests for the nodetypes?

**Comment by skurfuerst** [comment]
@dlubitz I do not have a clear preference here - so do what fits best for you :) 

All the best,
Sebastian

**Comment by dlubitz** [comment]
@skurfuerst I've added/changed some tests to cover the new feature. Luckily this revealed an issue with the NodeTypeName changes.

**Review by skurfuerst — APPROVED** [review]
really cool @dlubitz :) I think this can be merged, wdyt?

**Review by kitsunet — APPROVED** [review]
Seems useful to me!

**Pull request merged**